### ISSUE TITLE
BuildStepsExecutionException when running jibDockerBuild 

### ISSUE THREAD

**Issue by bwgjoseph** [GoogleContainerTools/jib#3740]
**Environment**:

- *Jib version:* 3.2.1 (jib-gradle)
- *Build tool:* gradle 7.5
- *OS:* Windows 10 x64
- *JDK:* temurin-jdk-17.0.4+8

**Description of the issue**:

When running the following command `./gradlew jibDockerBuild`, it throws the following error

```logs
BuildStepsExecutionException: unable to move: %APPDATA%\Local\Google\Jib\Cache\tmp\12345 to %APPDATA%\Local\Google\Jib\Cache\layers\54321; such failures are often caused by interference from antivirus (#3127), or rarely if the operation is not supported by the file system (for example: special non-local file system)
```

**Expected behavior**:

Running the command should build a local image using the local Docker Daemon

**Steps to reproduce**:

  1. run `./gradlew jibDockerBuild`

**`jib-gradle-plugin` Configuration**: 
```groovy
jib {
  from {
    image = 'eclipse-temurin:17.0.4_8-jdk'
  }
  to {
    image = 'my-image/built-with-jib'
    tags = ['tag2', 'latest']
  }
  container {
    jvmFlags = ['-Xms512m', '-Xdebug']
    ports = ['8080']
  }
}
```

**Log output**: 

**Additional Information**: 

The error message points me to this https://github.com/GoogleContainerTools/jib/issues/1635#issuecomment-484199610 where it did mentioned about anti-virus affecting it. My research also brings me to #3127 and #1438, similarly, my machine is installed with `mcafee endpoint security 10.7` but I have no control over it.

I have tried to run with the following command `./gradlew jibDockerBuild -Djib.applicationCache=D:\cache -Djib.seralize=true` based on some suggestion I saw over the different issues, but it does not seem to work as well.

**Note**: I should probably also mention that it used to work when I last ran it a few months back. I did a trace, but I don't see any major upgrade in between this period for `gradle/jib` but I'm not sure about anti-virus though.

I am unable to provide a full log as I'm working on an air-gapped environment. Let me know if anything else is required, and I will try to provide

Thanks

**Comment by chanseokoh** [comment]
The problematic directory (`%APPDATA%\Local\Google\Jib\Cache\tmp\12345`) is not an application cache directory. (An application cache is local and exclusive to your single project and resides under your Gradle build output directory (`<Gradle project root>/build`).) Therefore, setting `-Djib.applicationCache` won't make a difference.

The directory in question is the central base image cache shared across Jib projects on the same machine, so try `-Djib.baseImageCache=...` for testing. However, remember that this is changing the shared cache, so if you decide to do this for a short-term solution, you'll want to set this for all Jib runs across projects to benefit the cache.

I bet you won't need `-Djib.seralize=true`.

**Comment by bwgjoseph** [comment]
If I understand you correctly, the issue is with the central base image cache, and thus, by changing the directory, would probably solve my issue, right?

> However, remember that this is changing the shared cache, so if you decide to do this for a short-term solution, you'll want to set this for all Jib runs across projects to benefit the cache.

I don't quite understand this portion, so is there a long term solution for this? Is the issue caused by AV, and thus, unless I disable or switch AV, the only way out is to change the base image cache directory?

**Comment by chanseokoh** [comment]
> If I understand you correctly, the issue is with the central base image cache, and thus, by changing the directory, would probably solve my issue, right?

Yes. The assumption is that AV is opening and holding the directory newly created by Jib (maybe for scanning) for a while and preventing Jib from atomically moving (renaming) the directory. This issue is only specific to the Windows OS with some AV.

> is there a long term solution for this? Is the issue caused by AV, and thus, unless I disable or switch AV, the only way out is to change the base image cache directory?

Hopefully, you can change the settings of your AV to exclude the Jib base image cache from its scanning scope so that you don't have to keep setting `-Djib.baseImageCache=...` in the mean time. Disabling or switching AV is obviously an option, but yeah, if that's not possible, the only way out I can think of is to keep setting `-Djib.baseImageCache=...` for all your Jib executions.

**Comment by bwgjoseph** [comment]
> The assumption is that AV is opening and holding the directory newly created by Jib

Jib creates new directory every time?

> Hopefully, you can change the settings of your AV to exclude the Jib base image cache

Jib default base image cache directory is `%APPDATA%\Local\Google\Jib\`?

I will try out your suggestion, and report back if it works. Although, I can't quite get it why it worked previously, and not now. I can't tell anything major changes, unless it's AV version, which I am not aware of the version when it was working.

**Comment by chanseokoh** [comment]
The base image cache is to cache images, which is comprised of layers (tar files in reality) and some metadata files. So, whenever it needs to download and save new images or layers, those will be created under the cache directory with some hierarchy. A lot of directories and files will be created in that cache, as you keep downloading new images.

> Jib default base image cache directory is `%APPDATA%\Local\Google\Jib\`?

Basically, `%LOCALAPPDATA%\Google\Jib\Cache`, [by default](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#system-properties).

> ** The default base image cache is in the following locations on each platform:
> ...
> Windows: `[cache root]\Google\Jib\Cache`, where `[cache root]` is `%XDG_CACHE_HOME%` (`%LOCALAPPDATA%` if not set)

**Comment by chanseokoh** [comment]
1. This is out of Jib's control. It means any Java program on your JDK just cannot establish SSL connections to your server. See the following issues for some ideas to debug your environment:
   * https://github.com/GoogleContainerTools/jib/issues/2458#issuecomment-625839215
   * #2482
   * https://github.com/GoogleContainerTools/jib/issues/1966#issuecomment-529662786

   Note it's an unrelated issue.

2. I don't think `httpTimeout` helps for anything. The issue is that your system is blocking you from moving a file on your local disk. No network connection is involved.

3. It may be that your AV is also monitoring `D:\Jib`. The error message is what it is saying: for some reason, a trivial I/O operation to rename a directory is failing. [Here](https://github.com/GoogleContainerTools/jib/issues/1635#issuecomment-775407097), a user reported that
   > * I set the path under the c drive where I keep tooling caches such as the local maven repo
   > * I can't turn off the antivirus as I do not have rights :(
   > * the new location is potentially not treated the same by the antivirus software

**Comment by bwgjoseph** [comment]
1. Yup, it is an unrelated issue to this, but not sure why it thinks so though. I am using the same JDK for other various purpose, and it is running as it should. I will try to figure out again, don't want to pollute this issue with unrelated issue. 
2. I wasn't clear about it, it was because I didn't want the `GET` request to retry so long (20sec), so I tried to set a lower value but it didn't seem to take effect. Not sure if I understand the arg wrongly.
3. I guess I can only see if there's a way to temp disable the AV, or set to ignore certain directory to be able to test properly if it works and if the root cause is really the AV

I will not be able to test until a week or so later, will report back my findings again. 

Thanks for helping out! Appreciate it

**Comment by chanseokoh** [comment]
> Yup, it is an unrelated issue to this, but not sure why it thinks so though. 

Just to make it sure, "it" here is the SSL layer on your JDK, not Jib.

`jib.httpTimeout` is for connection and read timeout. Jib does [exponential back-off waiting on retying](https://github.com/GoogleContainerTools/jib/blob/3e694e7c7fcb3f56c333359ada9fb40450cf771c/jib-core/src/main/java/com/google/cloud/tools/jib/http/FailoverHttpClient.java#L360) on I/O exception in the hope that it's a transient error that may go away at some point. BTW, your error is not a recoverable one, so no matter how long Jib waits or how many times it retries, it will never succeed.

I wonder if `%SYSTEM%\WINDOWS\TEMP` or `%LOCALAPPDATA%\Temp` is worth trying.

**Comment by bwgjoseph** [comment]
Yup, that's correct on the SSL part.

That's a great idea to try on those directories, I shall do that and report back a week later. 

Thanks!

**Comment by bwgjoseph** [comment]
So it seems to worked by disabling On-Access Scan on my `McAfee Endpoint Security`

However, it did not work the first time when I used the `exclusion` feature. I ran using `gradlew jibDockerBuild -Djib.baseImageCache=D:\Jib\Cache -Djib.console=plain -Djib.allowInsecureRegistries=true`

The test I did was 

1. Exclude `D:\Jib` > run jibDockerBuild > didn't work
2. Disable on access scan > run jibDockerBuild > work
3. Enabled on access scan > didn't have any exclusion > run jibDockerBuild > work!

I also tried excluding `%LOCALAPPDATA%\Jib` and ran the command without specifying `baseImageCache` and the result is the same as above.

So it seems to me that once it has cached the first time for jib, it won't copy any files anymore? Thus, won't trigger the on access scan because no files are being copied?

I'm not sure why the exclusion didn't work, but is there other directory I should try excluding to test?

**Comment by chanseokoh** [comment]
Yes. Once Jib downloaded and cached images, Jib will not download them again, for the obvious reason.

> is there other directory I should try excluding to test?

If you remember, the exception message already shows the path in question.

**Comment by bwgjoseph** [comment]
> If you remember, the exception message already shows the path in question.

Yup, I added the directories as mentioned, but the exclusion didn't work. So was wondering if it could have other directory that I did not include. But if those are the only ones, then it might be something wrong with the exclusion feature from the AV, and I will treat it as such. 

Otherwise, I think we can close this issue, and thank you so much for your assistance. Appreciate it!
